File: packages/guards/src/answer-citation-needed.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface AnswerCitationNeededOptions {
  action: 'block' | 'warn';
  threshold?: number;
}

const UNCITED_PATTERNS: RegExp[] = [
  /\b\d{1,3}(?:,\d{3})*(?:\.\d+)?\s*%/,
  /\bstudies?\s+(?:show|suggest|indicate|found|demonstrate)\b/i,
  /\bresearch\s+(?:shows?|suggests?|indicates?|found|demonstrates?)\b/i,
  /\baccording\s+to\s+(?:experts?|scientists?|researchers?)\b/i,
  /\bstatistic(?:s|ally)\s/i,
  /\b(?:survey|poll|report)\s+(?:shows?|found|reveals?)\b/i,
  /\bscientific(?:ally)?\s+(?:proven|established)\b/i,
  /\b(?:data|evidence)\s+(?:shows?|suggests?|indicates?)\b/i,
  /\bin\s+\d{4}\s*,\s*(?:a\s+)?(?:study|report|survey)\b/i,
  /\b(?:approximately|roughly|about)\s+\d+(?:\.\d+)?\s*(?:million|billion|trillion|percent|%)\b/i,
];

const CITATION_MARKERS: RegExp[] = [
  /\[[\d,\s]+\]/,
  /\((?:(?:19|20)\d{2})\)/,
  /https?:\/\//,
  /(?:source|reference|citation|ref)\s*[:]/i,
];

function hasCitation(text: string): boolean {
  return CITATION_MARKERS.some((p) => p.test(text));
}

export function answerCitationNeeded(options: AnswerCitationNeededOptions): Guard {
  const threshold = options.threshold ?? 3;

  return {
    name: 'answer-citation-needed',
    version: '0.1.0',
    description: 'Flags claims that need citations',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      let uncited = 0;

      const sentences = text.split(/(?<=[.!?])\s+/);
      for (const s of sentences) {
        const needsCite = UNCITED_PATTERNS.some((p) => p.test(s));
        if (needsCite && !hasCitation(s)) uncited++;
      }

      const triggered = uncited >= threshold;
      const score = triggered ? Math.min(uncited / (threshold * 2), 1.0) : 0;

      return {
        guardName: 'answer-citation-needed',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { uncitedClaims: uncited, threshold } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/answer_citation_needed.py
"""Flag claims that need citations."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_UNCITED: list[re.Pattern[str]] = [
    re.compile(r"\b\d{1,3}(?:,\d{3})*(?:\.\d+)?\s*%"),
    re.compile(
        r"\bstudies?\s+(?:show|suggest|indicate"
        r"|found|demonstrate)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bresearch\s+(?:shows?|suggests?"
        r"|indicates?|found|demonstrates?)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\baccording\s+to\s+(?:experts?"
        r"|scientists?|researchers?)\b",
        re.IGNORECASE,
    ),
    re.compile(r"\bstatistic(?:s|ally)\s", re.IGNORECASE),
    re.compile(
        r"\b(?:survey|poll|report)"
        r"\s+(?:shows?|found|reveals?)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bscientific(?:ally)?\s+(?:proven|established)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:data|evidence)"
        r"\s+(?:shows?|suggests?|indicates?)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\bin\s+\d{4}\s*,\s*(?:a\s+)?"
        r"(?:study|report|survey)\b",
        re.IGNORECASE,
    ),
    re.compile(
        r"\b(?:approximately|roughly|about)"
        r"\s+\d+(?:\.\d+)?\s*"
        r"(?:million|billion|trillion|percent|%)\b",
        re.IGNORECASE,
    ),
]

_CITATION_MARKERS: list[re.Pattern[str]] = [
    re.compile(r"\[[\d,\s]+\]"),
    re.compile(r"\((?:(?:19|20)\d{2})\)"),
    re.compile(r"https?://"),
    re.compile(
        r"(?:source|reference|citation|ref)\s*[:]",
        re.IGNORECASE,
    ),
]


def _has_citation(text: str) -> bool:
    return any(p.search(text) for p in _CITATION_MARKERS)


class _AnswerCitationNeeded:
    def __init__(
        self, *, action: str = "warn", threshold: int = 3
    ) -> None:
        self.name = "answer-citation-needed"
        self.action = action
        self._threshold = threshold

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        uncited = 0

        sentences = re.split(r"(?<=[.!?])\s+", text)
        for s in sentences:
            needs = any(p.search(s) for p in _UNCITED)
            if needs and not _has_citation(s):
                uncited += 1

        triggered = uncited >= self._threshold
        score = (
            min(uncited / (self._threshold * 2), 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="answer-citation-needed",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Uncited claims detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "uncited_claims": uncited,
                    "threshold": self._threshold,
                }
                if triggered
                else None
            ),
        )


def answer_citation_needed(
    *, action: str = "warn", threshold: int = 3
) -> _AnswerCitationNeeded:
    return _AnswerCitationNeeded(
        action=action, threshold=threshold
    )
